components/raftstore/src/store/txn_ext.rs

```diff
@@ -47,7 +47,47 @@ impl fmt::Debug for TxnExt {
 
 /// Pessimistic locks of a region peer.
 pub struct PeerPessimisticLocks {
-    pub map: HashMap<Key, PessimisticLock>,
+    /// The table that stores pessimistic locks.
+    ///
+    /// The bool marks an ongoing write request (which has been sent to the raftstore while not
+    /// applied yet) will delete this lock. The lock will be really deleted after applying the
+    /// write request. The flag will decide whether this lock should be migrated to other peers
+    /// on leader or region changes:
+    ///
+    /// - Transfer leader
+    ///   The lock with the deleted mark SHOULD NOT be proposed before transferring leader.
+    ///   Considering the following cases with different orders:
+    ///   1. Propose write -> propose locks -> apply write -> apply locks -> transfer leader
+    ///      Because the locks marking deleted will not be proposed. The lock will be deleted when
+    ///      applying the write while not showing up again after applying the locks.
+    ///   2. Propose locks -> propose write -> transfer leader
+    ///      No lock will be lost in normal cases because the write request has been sent to the
+    ///      raftstore, it is likely to be proposed successfully, while the leader will need at
+    ///      least another round to receive the transfer leader message from the transferree.
+    ///  
+    /// - Split region
+    ///   The lock with the deleted mark SHOULD be moved to new regions on region split.
+    ///   Considering the following cases with different orders:
+    ///   1. Propose write -> propose split -> apply write -> execute split
+    ///      The write will be applied earlier than split. So, the lock will be deleted earlier
```

Comment:
Deleting the lock is executed in the scheduler worker pool. It may be after executing split.

Reply:
Ah, I didn't notice it. But this should be easy to solve. We can move deleting the lock to be executed in the apply pool.